lib/block_view_stake.go

```diff
@@ -0,0 +1,2476 @@
+
+	// Convert ValidatorPublicKey to ValidatorPublicKeyBase58Check.
+	validatorPublicKeyBase58Check := PkToString(metadata.ValidatorPublicKey.ToBytes(), bav.Params)
+
+	// Construct TxindexMetadata.
+	txindexMetadata := &UnstakeTxindexMetadata{
+		StakerPublicKeyBase58Check:    stakerPublicKeyBase58Check,
+		ValidatorPublicKeyBase58Check: validatorPublicKeyBase58Check,
+		UnstakeAmountNanos:            metadata.UnstakeAmountNanos,
+	}
+
+	// Construct AffectedPublicKeys.
+	affectedPublicKeys := []*AffectedPublicKey{
+		{
+			PublicKeyBase58Check: stakerPublicKeyBase58Check,
+			Metadata:             "UnstakerPublicKeyBase58Check",
+		},
+		{
+			PublicKeyBase58Check: validatorPublicKeyBase58Check,
+			Metadata:             "ValidatorUnstakedFromPublicKeyBase58Check",
+		},
+	}
+
+	return txindexMetadata, affectedPublicKeys
+}
+
+func (bav *UtxoView) CreateUnlockStakeTxindexMetadata(utxoOp *UtxoOperation, txn *MsgDeSoTxn) (*UnlockStakeTxindexMetadata, []*AffectedPublicKey) {
+	metadata := txn.TxnMeta.(*UnlockStakeMetadata)
+
+	// Convert TransactorPublicKeyBytes to StakerPublicKeyBase58Check.
+	stakerPublicKeyBase58Check := PkToString(txn.PublicKey, bav.Params)
+
+	// Convert ValidatorPublicKey to ValidatorPublicKeyBase58Check.
+	validatorPublicKeyBase58Check := PkToString(metadata.ValidatorPublicKey.ToBytes(), bav.Params)
+
+	// Calculate TotalUnlockedAmountNanos.
+	totalUnlockedAmountNanos := uint256.NewInt()
+	var err error
+	for _, prevLockedStakeEntry := range utxoOp.PrevLockedStakeEntries {
+		totalUnlockedAmountNanos, err = SafeUint256().Add(
+			totalUnlockedAmountNanos, prevLockedStakeEntry.LockedAmountNanos,
+		)
+		if err != nil {
+			glog.Errorf("CreateUnlockStakeTxindexMetadata: error calculating TotalUnlockedAmountNanos: %v", err)
+			totalUnlockedAmountNanos = uint256.NewInt()
+			break
+		}
+	}
+
+	// Construct TxindexMetadata.
+	txindexMetadata := &UnlockStakeTxindexMetadata{
+		StakerPublicKeyBase58Check:    stakerPublicKeyBase58Check,
+		ValidatorPublicKeyBase58Check: validatorPublicKeyBase58Check,
+		StartEpochNumber:              metadata.StartEpochNumber,
+		EndEpochNumber:                metadata.EndEpochNumber,
+		TotalUnlockedAmountNanos:      totalUnlockedAmountNanos,
+	}
+
+	// Construct AffectedPublicKeys.
+	affectedPublicKeys := []*AffectedPublicKey{
+		{
+			PublicKeyBase58Check: stakerPublicKeyBase58Check,
+			Metadata:             "UnlockedStakerPublicKeyBase58Check",
+		},
+	}
+
+	return txindexMetadata, affectedPublicKeys
+}
+
+//
+// TRANSACTION SPENDING LIMITS
+//
+
+type StakeLimitKey struct {
+	ValidatorPKID PKID
+	StakerPKID    PKID
+}
+
+func MakeStakeLimitKey(validatorPKID *PKID, stakerPKID *PKID) StakeLimitKey {
+	return StakeLimitKey{
+		ValidatorPKID: *validatorPKID,
+		StakerPKID:    *stakerPKID,
+	}
+}
+
+func (stakeLimitKey *StakeLimitKey) Encode() []byte {
+	var data []byte
+	data = append(data, stakeLimitKey.ValidatorPKID.ToBytes()...)
```

Comment:
I didn't use `EncodeToBytes` here only because we don't have the blockHeight + skipMetadata fields handy here to use. FWIW, this is what we also did for associations limit key [here](https://github.com/deso-protocol/core/blob/main/lib/network.go#L6655). Happy to modify though if you think this will be an issue.